docs/decisions/0006-pluggable-lti-configuration.rst

```diff
@@ -0,0 +1,128 @@
+Pluggable and re-usable LTI configuration
+-----------------------------------------
+
+
+Status
+======
+
+Proposal
+
+Context
+=======
+
+LTI Tools provide extra funcionality and allows course creators to embed external interactive content
+in the LMS through a standardized setup.
+
+Currently in the Open edX platform:
+
+#. Each LTI tool needs to be configured individually, sometimes requiring multiple valiables to be set to get
+a working integration. This makes the course creation process cumbersome on courses that heavily rely on LTI 
+components. 
+#. There's no support for instance-wide LTI tool configuration or any sort of pluggable LTI configurations.
+
+
+This proposal aims to solve these issues by introducing a mechanism to allow LTI configuration pluggability 
+and changes to the XBlock to allow re-using configurations to reduce instructor friction.
+
+
+Decision
+========
+
+The decision is to enable LTI reusability and pluggability through a combination of a filter hook (as defined in `OEP-50`_) 
+and a standardized API interface. 
+Through the API interface, plugins should be able to set up externally managed LTI configurations store their data, while
+the LTI consumer handles displaying the content.
+
+To enable this, the LTI Configuration model needs to be modified to:
+
+#. Allow externally configured tools (a new {{CONFIG_EXTERNAL}} choice for the configuration storage).
+#. Two new fields to store an unique external tool id and the plugin name/location.
+#. Changes to the {{get_lti_consumer}} methods to allow instancing a {{LtiConsumer}} from externally managed configurations.
+
+
+The filter part will allow the XBlock LTI Consumer to query re-usable and/or plugabble LTI configurations and display them to course 
+creators as configuration options in Studio.
+
+The filter response should contain all the necessary information for the XBlock to locate and use the externally managed LTI configuration.
+The reponse **should** be structured this way:
+
+.. code-block:: json
+    {
+        "filter_name": {
+            "api_path": "plugin_1_module_name.api.get_lti_config",
+            "available_configurations": [
+                {
+                    "lti_version": "LTI_1P1",
+                    "display_name": "Reusable LTI tool 1",
+                    "lti_config_id": 2,
+                }
+                {
+                    "lti_version": "LTI_1P3",
+                    "display_name": "Reusable LTI tool 2 - LTI 1.3",
+                    "lti_config_id": 3,     
+                }
+            ]
+        }
+    }
```

Comment:
I'm not sure I'm understanding this correctly. What if the user has a number of different filters installed? For instance, a filter that fetches the config from the database, another from the settings, another from a vault service? 

I think for LTI 1.1 we have a pretty straightforward mechanism where we can just have the filter directly return the two string values. We should use a simpler filter for that where each filter can simply go through all the existing config options, remove them, and add new ones. 

For LTI 1.3 I guess things are more complex, but what the filter can provide are raw config values that are then copied over to the block/database, after that the block work exactly the same as now.

Reply:
@xitij2000 I need to improve the wording here.

I though of "pluggable LTI configurations" as a combination of two things: 
1. One filter pipeline that solely exists to retrieve information about where to fetch a given LTI configuration + some metadata to display to the end user.
2. A standard python API that the xblock-lti-consumer would call to retrieve the actual configuration when LTI launches happened. We can replace the standard API with a second (different) filter pipeline that returns the LTI configuration given a UUID.
  
>What if the user has a number of different filters installed? For instance, a filter that fetches the config from the database, another from the settings, another from a vault service?

The result of the filter will simply contain more results (one for each LTI configuration available). This filter isn't supposed to return the LTI credentials/configuration, but the list of available configurations and how to retrieve data from them.

>For LTI 1.3 I guess things are more complex, but what the filter can provide are raw config values that are then copied over to the block/database, after that the block work exactly the same as now.

:+1: for the filter to provide settings, but not for them to be stored in the database as it wouldn't fulfill the client's requirements.